### PR TITLE
Allow gc to run in parallel on cache drives

### PR DESCRIPTION
**Pull request by poornas**
## Motivation and Context
Since list of entries qualifying for eviction are bounded with low and high watermark and will occupy constant memory, it is ok to run the gc in parallel across cache drives. Hence removing the lock held across drives here.  Per discussion with @abperiasamy 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#01336ca1-e3f6-42d3-9a54-a4a6b4219cf4) |
| mint-gateway-azure.sh | :x: [more...](#cf593540-98ec-449f-94d6-0ec30c40d8fc) |
| mint-gateway-nas.sh | :x: [more...](#57d31cad-e63f-4449-8848-18f3cf3e6497) |
<a name="cf593540-98ec-449f-94d6-0ec30c40d8fc"></a>
### 9051-f14ab8d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.58:30661
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a8e33720faa4:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 19 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... FAILED in 3 seconds
{
  "name": "awscli",
  "duration": 1542,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 14 tests successfully.
```
<a name="57d31cad-e63f-4449-8848-18f3cf3e6497"></a>
### 9051-f14ab8d/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30074
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2ef4275faa16:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-3xdcyqbj5jsadb",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120386,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:30074/aws-sdk-go-test-3xdcyqbj5jsadb: dial tcp 72.28.97.55:30074: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="01336ca1-e3f6-42d3-9a54-a4a6b4219cf4"></a>
### 9051-f14ab8d/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30583
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 07a7e3615d3c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-oz2nujozs2sev0",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365103,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15F7156915DE2225, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#519ef561-2166-4427-a6ff-4dc8ef505fd8) |
| mint-gateway-azure.sh | :x: [more...](#2a387e3f-53c8-4d9f-89e4-66541d44eccd) |
<a name="2a387e3f-53c8-4d9f-89e4-66541d44eccd"></a>
### 9051-1c20bdc/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31921
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2f0ec07816e3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 15 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... FAILED in 2 minutes and 49 seconds
{
  "name": "awscli",
  "duration": 1531,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 14 tests successfully.
```
<a name="519ef561-2166-4427-a6ff-4dc8ef505fd8"></a>
### 9051-1c20bdc/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31039
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e14f4bcdc3b0:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 17 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 8 seconds
{
  "name": "mc",
  "duration": "3150",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-524 --quiet --no-color cp myminio/mc-test-bucket-19108/mc-test-object-2202 mc-test-object-2202.downloaded  >>> \n`myminio/mc-test-bucket-19108/mc-test-object-2202` -> `mc-test-object-2202.downloaded`\nmc: <ERROR> Failed to copy `http://72.28.97.56:31039/mc-test-bucket-19108/mc-test-object-2202`. read tcp 172.17.0.2:33986->72.28.97.56:31039: read: connection reset by peer\nTotal: 0 B, Transferred: 24.09 MiB, Speed: 29.17 MiB/s"
}

Executed 6 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by klauspost** [comment]
CI failure is unrelated.

File: cmd/disk-cache.go [hunk 1]
```diff
@@ -711,14 +710,12 @@ func (c *cacheObjects) gc(ctx context.Context, doneCh chan struct{}) {
 				if dcache.gcCount() == 0 {
 					continue
 				}
-				gcLock.Lock()
 				wg.Add(1)
-				go func(d *diskCache, l *sync.Mutex) {
+				go func(d *diskCache) {
```
Comment (klauspost) [inline cmd/disk-cache.go:714]: Wait. I thought this would only start if we reached the high watermark?

Comment (poornas) [reply cmd/disk-cache.go:714]: that is correct @klauspost - line 710 checks for this. When high watermark is reached, the gcCounter gets incremented for each cache access until gc is started on that drive.

**Review by klauspost — APPROVED** [review]
LGTM.

**Pull request merged**